File: src/lib/color.ts
function hslToHex(h: number, s: number, l: number): string {
	l /= 100;
	s /= 100;

	const k = (n: number) => (n + h / 30) % 12;
	const a = s * Math.min(l, 1 - l);
	const f = (n: number) => l - a * Math.max(-1, Math.min(k(n) - 3, Math.min(9 - k(n), 1)));

	const r = Math.round(255 * f(0));
	const g = Math.round(255 * f(8));
	const b = Math.round(255 * f(4));

	const toHex = (x: number) => x.toString(16).padStart(2, '0');

	return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
}

export function generateSoftColor() {
	const h = Math.floor(Math.random() * 360); // Any hue
	const s = Math.floor(Math.random() * 30) + 30; // 30–60% saturation
	const l = Math.floor(Math.random() * 20) + 75; // 75–95% lightness

	return hslToHex(h, s, l);
}

File: beer-api/src/utils/color.rs
use rand::RngExt;

pub(crate) fn generate_soft_color() -> String {
    let mut rng = rand::rng();
    hsl_to_hex(
        f64::from(rng.random_range(0..360)),
        f64::from(rng.random_range(30..60)),
        f64::from(rng.random_range(75..95)),
    )
}

fn hsl_to_hex(hue: f64, saturation: f64, lightness: f64) -> String {
    let saturation = saturation / 100.0;
    let lightness = lightness / 100.0;
    let a = saturation * lightness.min(1.0 - lightness);
    #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
    let component = |n: f64| {
        let k = (n + hue / 30.0) % 12.0;
        (255.0 * (lightness - a * (k - 3.0).min(9.0 - k).clamp(-1.0, 1.0))).round() as u8
    };
    format!(
        "#{:02x}{:02x}{:02x}",
        component(0.0),
        component(8.0),
        component(4.0)
    )
}
